src/VisualStudio/CodeLens/ReferenceCodeLensProvider.cs

```diff
@@ -0,0 +1,293 @@
+
+            public DataPoint(
+                ReferenceCodeLensProvider owner, CodeLensDescriptor descriptor, Stream stream, ICodeLensCallbackService callbackService)
+            {
+                _owner = owner;
+                this.Descriptor = descriptor;
+
+                _roslynRpc = new JsonRpc(new JsonRpcMessageHandler(stream, stream), target: new RoslynCallbackTarget(Invalidate));
+                _roslynRpc.JsonSerializer.Converters.Add(AggregateJsonConverter.Instance);
+
+                _roslynRpc.StartListening();
+
+                _callbackService = callbackService;
+            }
+
+            public event AsyncEventHandler InvalidatedAsync;
+
+            public CodeLensDescriptor Descriptor { get; }
+
+            public async Task<CodeLensDataPointDescriptor> GetDataAsync(CodeLensDescriptorContext descriptorContext, CancellationToken cancellationToken)
+            {
+                // we always get data through VS rather than Roslyn OOP directly since we want final data rather than
+                // raw data from Roslyn OOP such as razor find all reference results
+                var referenceCount = await _callbackService.InvokeAsync<ReferenceCount>(
+                    _owner, nameof(ICodeLensContext.GetReferenceCountAsync), new object[] { this.Descriptor, descriptorContext }, cancellationToken).ConfigureAwait(false);
+
+                if (referenceCount == null)
+                {
+                    return null;
+                }
+
+                var referenceCountString = $"{referenceCount.Count}{(referenceCount.IsCapped ? "+" : string.Empty)}";
+
+                return new CodeLensDataPointDescriptor()
+                {
+                    Description = referenceCount.Count == 1
+                        ? string.Format(CodeLensVSResources._0_reference, referenceCountString)
+                        : string.Format(CodeLensVSResources._0_references, referenceCountString),
+                    IntValue = referenceCount.Count,
+                    TooltipText = string.Format(CodeLensVSResources.This_0_has_1_references, GetCodeElementKindsString(Descriptor.Kind), referenceCountString),
+                    ImageId = null
+                };
+
+                string GetCodeElementKindsString(CodeElementKinds kind)
+                {
+                    switch (kind)
+                    {
+                        case CodeElementKinds.Method:
+                            return CodeLensVSResources.method;
+                        case CodeElementKinds.Type:
+                            return CodeLensVSResources.type;
+                        case CodeElementKinds.Property:
+                            return CodeLensVSResources.property;
+                        default:
+                            // code lens engine will catch and ignore exception
+                            // basically not showing data point
+                            throw new NotSupportedException(nameof(kind));
```

Comment:
Can we call this method first, so we don't waste time to do all the work then ended up throwing?